### PR TITLE
Protect curriculum example projects from report abuse spam 

### PR DESCRIPTION
**Pull request by Erin007**
**[LP-1568](https://codedotorg.atlassian.net/browse/LP-1568): "Adjust abuse scores so it would take MANY students to report these as abusive."**
The Curriculum team [reported via Slack ](https://codedotorg.slack.com/archives/CA3KCSGTD/p1600725916019600) a run of spammy abuse reports on internally-made projects that are used as examples in CSD, which was causing the projects to be flagged for abuse. The Curriculum team was able to put the projects back in working order by using the admin panel to reset abuse scores to 0.  However, there was a request to prevent this from happening in the future by buffering abuse scores.  There's a trick for this; by featuring and un-featuring a project [you can set the project's abuse score to -50](https://github.com/code-dot-org/code-dot-org/blob/ee7ee63728e084d1558430f1ae4cc0945bc8b3f7/dashboard/app/controllers/featured_projects_controller.rb#L9). We shouldn't rely on this sneaky bit of insider knowledge, though, so I made the process of buffering  an abuse score for a project much more obvious from the admin panel.  I also made the reset abuse link a button so both options would be more visible, and made the abuse score information visible even if the project isn't abusive.
 
buffer-abuse-button

**[LP-1569](https://codedotorg.atlassian.net/browse/LP-1569): "Any projects created by the curriculum team's Coursedocs account will not be able to be reported for abuse"**  

This PR also implements a broader, long-term solution that doesn't depend on someone buffering or resetting the abuse score for projects in the future. I did this by preventing abuse reports from changing the score of projects created by users who have project validator permissions. Next, I will give the coursedocs@code.org account project validator permissions, which will mean any project created by the curriculum team under that account will not be able to be reported for abuse.

### CODE REVIEW

File: dashboard/app/controllers/report_abuse_controller.rb [hunk 1]
```diff
@@ -19,70 +19,81 @@ def to_honeybadger_context
 class ReportAbuseController < ApplicationController
   AGE_CUSTOM_FIELD_ID = 24_024_923
 
+  # Projects that are created by users with project validator permissions are
+  # blocked from abuse reports because they are created internally and we know # they are safe. This reduces spamming of the report abuse feature.
```
Comment (clareconstantine) [inline dashboard/app/controllers/report_abuse_controller.rb]: super tiny, looks like this was intended to be separate lines.

**Review by clareconstantine — APPROVED** [review]
Thanks for jumping on this, Erin, and for making this easy to do in the future. Looks awesome!

File: dashboard/app/controllers/report_abuse_controller.rb [hunk 2]
```diff
@@ -19,70 +19,82 @@ def to_honeybadger_context
...
   AGE_CUSTOM_FIELD_ID = 24_024_923
 
+  # Projects that are created by users with project validator permissions are
+  # blocked from abuse reports because they are created internally and we know
+  # they are safe. This reduces spamming of the report abuse feature.
+  def protected_project?
+    return false if params[:channel_id].blank?
+    owner, _storage_app_id = storage_decrypt_channel_id(params[:channel_id])
+    project_owner = User.find(owner)
+    project_owner.permission?(UserPermission::PROJECT_VALIDATOR)
```
Comment (davidsbailey) [inline dashboard/app/controllers/report_abuse_controller.rb]: nice approach! what have you done to validate that this is working?

hint: I don't think it is possible for this to be working reliably without looking at the user_storage_ids table.

File: dashboard/app/controllers/report_abuse_controller.rb [hunk 3]
```diff
@@ -19,70 +19,82 @@ def to_honeybadger_context
...
-            custom_fields: [{id: AGE_CUSTOM_FIELD_ID, value: params[:age]}],
-            tags: (params[:abuse_type] == 'infringement' ? ['report_abuse', 'infringement'] : ['report_abuse'])
-          }
-        }.to_json,
-        basic_auth: {username: 'dev@code.org/token', password: Dashboard::Application.config.zendesk_dev_token}
-      )
-      raise ZendeskError.new(response.code, response.body) unless response.success?
-    end
+    unless protected_project?
+      unless Rails.env.development? || Rails.env.test?
```
Comment (davidsbailey) [inline dashboard/app/controllers/report_abuse_controller.rb:36]: why is this needed? we try to keep these kinds of checks out of our application code because it makes them less testable. if you do need to perform this check, please consider using a CDO variable which can then be stubbed during testing.

**Review by davidsbailey — COMMENTED** [review]
nice solutions Erin! Some comments inline.

File: dashboard/app/controllers/report_abuse_controller.rb (hunk 3 above)
Comment (Erin007) [reply dashboard/app/controllers/report_abuse_controller.rb:36]: Do you mean why is `unless Rails.env.development? || Rails.env.test?`needed? It's there so we don't spam Zendesk with tickets from test and development.  I don't think I understand what you're suggesting.

Comment (davidsbailey) [reply dashboard/app/controllers/report_abuse_controller.rb:36]: sorry, I missed that this check was in there already. I guess it is fine to leave, then.

I was suggesting defining a variable like `CDO.send_abuse_reports` which would be true in production and false in development / test.

Comment (Erin007) [reply dashboard/app/controllers/report_abuse_controller.rb:36]: Oh, ok.  Got it. Thanks for clarifying. That makes sense, now.

File: dashboard/app/controllers/report_abuse_controller.rb (hunk 2 above)
Comment (Erin007) [reply dashboard/app/controllers/report_abuse_controller.rb]: 🤦‍♀️ Good catch! I updated and added test coverage to verify.

File: dashboard/app/controllers/report_abuse_controller.rb [hunk 4]
```diff
@@ -19,70 +19,83 @@ def to_honeybadger_context
...
 
+  # Projects that are created by users with project validator permissions are
+  # blocked from abuse reports because they are created internally and we know
+  # they are safe. This reduces spamming of the report abuse feature.
+  def protected_project?
+    return false if params[:channel_id].blank?
+    owner_storage_id, _ = storage_decrypt_channel_id(params[:channel_id])
+    owner_user_id = user_storage_ids_table.where(id: owner_storage_id).first[:user_id]
+    project_owner = User.find(owner_user_id)
+    project_owner.permission?(UserPermission::PROJECT_VALIDATOR)
```
Comment (davidsbailey) [inline dashboard/app/controllers/report_abuse_controller.rb:31]: looks better! now this looks right for projects owned by users, but may give an error for projects which do not appear in user_storage_ids, i.e. those created by signed-out users.

**Review by davidsbailey — APPROVED** [review]
LGTM after addressing last comment. nice improvements!
